annotated.go

```diff
@@ -89,3 +90,228 @@ func (a Annotated) String() string {
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If there are more Annotations are specified than than one per
+// Annotation type, none of the Annotations will be applied.
+//
+// For example,
+//
+// fx.Provide(
+//   fx.Annotate(
+//     NewGateWay,
+//     fx.ParamTags(`name:"ro" optional:"true"`),
+//     fx.ParamTags(`name:"rw"),
+//     fx.ResultTags(`name:"foo"`)
+//   )
+// )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+
+	if !verifyAnnotation(numIn, numOut, anns...) {
+		return f
+	}
+
+	var ins []reflect.Type
+	var outs []reflect.Type
+
+	annotatedIn := false
+	annotatedOut := false
+
+	for _, ann := range anns {
+		if pTags, ok := ann.(paramTags); ok {
+			ins = pTags.getAnnotatedType(fType)
+			annotatedIn = true
+		}
+		if rTags, ok := ann.(resultTags); ok {
+			outs = rTags.getAnnotatedType(fType)
+			annotatedOut = true
+		}
+	}
+
+	if !annotatedIn {
+		ins = make([]reflect.Type, numIn)
+		for i := 0; i < fType.NumIn(); i++ {
+			ins[i] = fType.In(i)
+		}
+	}
+	if !annotatedOut {
+		outs = make([]reflect.Type, numOut)
+		for i := 0; i < fType.NumOut(); i++ {
+			outs[i] = fType.Out(i)
+		}
+	}
+
+	var newF func([]reflect.Value) []reflect.Value
+
+	if annotatedIn && annotatedOut {
+		newF = func(args []reflect.Value) []reflect.Value {
+			fParams := make([]reflect.Value, numIn)
+			params := args[0]
+			for i := 0; i < numIn; i++ {
+				fParams[i] = params.Field(i + 1)
+			}
+			fResults := fVal.Call(fParams)
+			results := reflect.New(outs[0]).Elem()
+			for i := 0; i < numOut; i++ {
+				results.FieldByName(fmt.Sprintf("Field%d", i)).Set(fResults[i])
+			}
+			return []reflect.Value{results}
+		}
+	} else if annotatedIn {
+		newF = func(args []reflect.Value) []reflect.Value {
+			fParams := make([]reflect.Value, numIn)
+			params := args[0]
+			for i := 0; i < numIn; i++ {
+				fParams[i] = params.Field(i + 1)
+			}
+			return fVal.Call(fParams)
+		}
+	} else { // annotatedOut
```

Comment:
I think annotatedOut might want to guard against `error` results. There can be an extraneous error argument at the end which should not be in the `fx.Out`.